IPC testing API is unable to reply with nullopt optional

kkinnunen-apple opened the pull request:
#### 8ebae84b5572e076d84b8ead18258c48a72e9280
<pre>
IPC testing API is unable to reply with nullopt optional
<a href="https://bugs.webkit.org/show_bug.cgi?id=275409">https://bugs.webkit.org/show_bug.cgi?id=275409</a>
<a href="https://rdar.apple.com/129669346">rdar://129669346</a>

Reviewed by Cameron McCormack.

Decoding std::optional would return empty JSValue (JSValue()) if the
optional was std::nullopt. This would be interpreted as a decoding
failure.

Fix this by returning the more natural value, jsUndefined value. Fix the
result array construction to not use jsUndefined to signify that the
value should be a array buffer containing the data. Use empty JSValue to
signify this. Use exception scope to detect when JS exception would
cancel the decoding as a decoding failure.

Decoding failures would use the already cleared exception as the result
JS object, causing a crash.

Fix by first constructing the JS value, then clearing the exception from
the scope.

Add decoder for WebCore::ExceptionData, to be used in testing the
message this problem was noted.

* LayoutTests/ipc/async-with-reply-optional.html: Added.
* Source/WebKit/Platform/IPC/JSIPCBinding.cpp:
(IPC::jsValueForDecodedArgumentValue):
(IPC::putJSValueForDecodedArgumentAtIndexOrArrayBufferIfUndefined): Deleted.
* Source/WebKit/Platform/IPC/JSIPCBinding.h:
(IPC::jsValueForDecodedArgumentValue):
(IPC::putJSValueForDecodeArgumentInArray):
* Source/WebKit/Shared/IPCTester.cpp:
(WebKit::IPCTester::asyncOptionalExceptionData):
* Source/WebKit/Shared/IPCTester.h:
* Source/WebKit/Shared/IPCTester.messages.in:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::jsSendWithAsyncReply):

Canonical link: <a href="https://commits.webkit.org/279999@main">https://commits.webkit.org/279999@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a57230c6dea695abb2d30370adc2040879a8f70

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55489 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34812 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7955 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58475 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5921 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42433 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6116 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/44684 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/4059 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57517 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/32720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/47820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/25811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/29505 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/5148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4065 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51435 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/5416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60065 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30643 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/5540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/52117 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31727 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/47890 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/51579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32810 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8174 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31476 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam reviewed (approved):
In the commit message:

> Decoding std::optional would return JSValue if the optional was
std::nullopt.

Is there a word missing?

kkinnunen-apple committed:
IPC testing API is unable to reply with nullopt optional
https://bugs.webkit.org/show_bug.cgi?id=275409
rdar://129669346

Reviewed by Cameron McCormack.

Decoding std::optional would return empty JSValue (JSValue()) if the
optional was std::nullopt. This would be interpreted as a decoding
failure.

Fix this by returning the more natural value, jsUndefined value. Fix the
result array construction to not use jsUndefined to signify that the
value should be a array buffer containing the data. Use empty JSValue to
signify this. Use exception scope to detect when JS exception would
cancel the decoding as a decoding failure.

Decoding failures would use the already cleared exception as the result
JS object, causing a crash.

Fix by first constructing the JS value, then clearing the exception from
the scope.

Add decoder for WebCore::ExceptionData, to be used in testing the
message this problem was noted.

* LayoutTests/ipc/async-with-reply-optional.html: Added.
* Source/WebKit/Platform/IPC/JSIPCBinding.cpp:
(IPC::jsValueForDecodedArgumentValue):
(IPC::putJSValueForDecodedArgumentAtIndexOrArrayBufferIfUndefined): Deleted.
* Source/WebKit/Platform/IPC/JSIPCBinding.h:
(IPC::jsValueForDecodedArgumentValue):
(IPC::putJSValueForDecodeArgumentInArray):
* Source/WebKit/Shared/IPCTester.cpp:
(WebKit::IPCTester::asyncOptionalExceptionData):
* Source/WebKit/Shared/IPCTester.h:
* Source/WebKit/Shared/IPCTester.messages.in:
* Source/WebKit/WebProcess/WebPage/IPCTestingAPI.cpp:
(WebKit::IPCTestingAPI::jsSendWithAsyncReply):

Canonical link: https://commits.webkit.org/279999@main
